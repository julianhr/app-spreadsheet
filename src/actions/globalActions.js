import { createAction } from 'redux-starter-kit'


export const setActiveCell = createAction('SET_ACTIVE_CELL')
export const displayInputter = createAction('DISPLAY_INPUTTER')
