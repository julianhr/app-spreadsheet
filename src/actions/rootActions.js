import { createAction } from 'redux-starter-kit'


export const setActiveCell = createAction('SET_ACTIVE_CELL')

export function moveActiveCell(key) {
  return (dispatch, getState) => {
    const { activeCell, rows, columns } = getState()
    const currRow = activeCell.codePointAt(0) - 'a'.codePointAt(0)
    const currCol = Number.parseInt(activeCell[1]) - 1
    let nextRow = currRow
    let nextCol = currCol

    if (key === 'ArrowUp') { nextRow = currRow - 1 }
    else if (key === 'ArrowRight') { nextCol = currCol + 1 }
    else if (key === 'ArrowDown') { nextRow = currRow + 1 }
    else if (key === 'ArrowLeft') { nextCol = currCol - 1 }

    if (Math.min(nextRow, nextCol) >= 0 && nextRow < rows && nextCol < columns) {
      nextRow = String.fromCodePoint('a'.codePointAt(0) + nextRow)
      nextCol = nextCol + 1
      dispatch(setActiveCell(`${nextRow}${nextCol}`))
    }
  }
}
