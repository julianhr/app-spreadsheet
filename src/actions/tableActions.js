import graph from '~/formulas/graph'
import Interpreter from '../formulas/Interpreter'


export function setCellData(location, entered) {
  const interpreter = new Interpreter(location)
  const result = interpreter.interpret(entered)


  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function replaceCellData(location, entered, result) {
  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function clearCellData(location) {
  return (dispatch, getState) => {
    const value = getState().table[location]

    graph.delVertex(location)

    if (value === undefined) {
      return Promise.resolve()
    }


    return Promise.resolve(dispatch({
      type: 'DELETE_CELL_DATA',
      payload: location,
    }))
  }
}
