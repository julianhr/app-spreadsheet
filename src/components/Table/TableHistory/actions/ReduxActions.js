import reduxConnect from '../reduxConnect'


class ReduxActions {
  static reduxConnect = reduxConnect

  get reduxConnect() {
    return ReduxActions.reduxConnect
  }

  getCellData(location) {
    return this.reduxConnect.getState().tableData[location]
  }
}

export default ReduxActions
