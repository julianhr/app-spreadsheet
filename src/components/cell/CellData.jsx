import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { interpret } from 'xstate'

import displayMachine from './displayMachine'
import InputData from './InputData'
import EvaluatedData from './EvaluatedData'


const Wrapper = styled.div`
  cursor: cell;
  width: 100%;
  height: 100%;
  font-size: 13px;
  border-right: 1px solid #dfdfdf;
  border-bottom: 1px solid #dfdfdf;
`
class CellData extends React.PureComponent {
  static propTypes = {
    location: PropTypes.string.isRequired,
  }

  constructor() {
    super()
    this.evaluatedOnDoubleClick = this.evaluatedOnDoubleClick.bind(this)
    this.evaluatedOnKeyDownEditable = this.evaluatedOnKeyDownEditable.bind(this)
    this.inputOnEscape = this.inputOnEscape.bind(this)
    this.inputOnCommit = this.inputOnCommit.bind(this)
  }

  state = {
    displayState: displayMachine.initialState,
  }

  displayService = interpret(displayMachine)
    .onTransition(displayState => this.setState({ displayState }))

  componentDidMount() {
    this.displayService.start()
  }
  
  componentWillUnmount() {
    this.displayService.stop()
  }
  
  evaluatedOnDoubleClick() {
    this.displayService.send('EDITABLE_MODIFY')
  }

  evaluatedOnKeyDownEditable() {
    this.displayService.send('EDITABLE_REPLACE')
  }

  inputOnEscape() {
    this.displayService.send('STATIC_FOCUSED')
  }

  inputOnCommit() {
    this.displayService.send('STATIC')
  }

  renderData() {
    const { displayState } = this.state

    if (displayState.matches('editable')) {
      return (
        <InputData
          replaceValue={displayState.matches('editable.replace')}
          location={this.props.location}
          onEscape={this.inputOnEscape}
          onCommit={this.inputOnCommit}
        />
      )
    } else {
      return (
        <EvaluatedData
          location={this.props.location}
          onDoubleClick={this.evaluatedOnDoubleClick}
          onKeyDownEditable={this.evaluatedOnKeyDownEditable}
          isFocused={displayState.matches('static.focused')}
        />
      )
    }
  }

  render() {
    return (
      <Wrapper>
        {this.renderData()}
      </Wrapper>
    )
  }
}

export default CellData
