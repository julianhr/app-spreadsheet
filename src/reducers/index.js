import { configureStore } from 'redux-starter-kit'

import globalReducer from './globalReducer'
import tableReducer from './tableReducer'


const appStoreFn = () => (
  configureStore({
    reducer: {
      global: globalReducer,
      table: tableReducer,
    }
  })
)


export default appStoreFn
