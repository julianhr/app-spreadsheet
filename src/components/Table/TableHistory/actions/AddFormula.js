import ReduxActions from './ReduxActions'


class AddFormula extends ReduxActions {
  constructor(location, oldVertex, newVertex) {
    super()
    this.location = location
    this.entered = newVertex.entered
    this.result = newVertex.result
    this.previous = oldVertex && {
      entered: oldVertex.entered,
      result: oldVertex.result,
    }
  }

  execute() {
    this.reduxConnect.fn.setCellData(this.location, this.entered, this.result)
    this.reduxConnect.fn.setActiveCell(this.location)
  }

  unexecute() {
    if (this.previous) {
      const { entered, result } = this.previous
      this.reduxConnect.fn.setCellData(this.location, entered, result)
    } else {
      this.reduxConnect.fn.clearCellData(this.location)
    }

    this.reduxConnect.fn.setActiveCell(this.location)
  }
}

export default AddFormula
