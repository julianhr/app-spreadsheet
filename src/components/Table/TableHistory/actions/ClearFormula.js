import ReduxActions from './ReduxActions'


class ClearFormula extends ReduxActions {
  constructor(location) {
    super()
    this.previous = this.getCellData(location)
    this.location = location
  }

  execute() {
    this.reduxConnect.fn.clearCellData(this.location)
    this.reduxConnect.fn.setActiveCell(this.location)
  }

  unexecute() {
    if (!this.previous) { return }
    const { entered, result } = this.previous
    this.reduxConnect.fn.setCellData(this.location, entered, result)
    this.reduxConnect.fn.setActiveCell(this.location)
  }
}

export default ClearFormula