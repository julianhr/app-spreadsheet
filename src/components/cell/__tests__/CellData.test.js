import React from 'react'
import { create } from 'react-test-renderer'
import { render, fireEvent, cleanup } from 'react-testing-library'
import * as xstate from 'xstate'

import { setCellData } from '~/actions/tableActions'
import { appStoreGen } from '~/reducers'
import MockApp from '~/__tests__/__mocks__/MockApp'
import CellData from '../CellData'


const requiredProps = {
  location: 'B-2',
}

const testProps = {
  ...requiredProps,
}

function createApp(props) {
  const appStore = appStoreGen()
  const wrapper = create(
    <MockApp customStore={appStore}>
      <CellData {...props} />
    </MockApp>
  )
  const instance = wrapper.root.findByType(CellData).instance
  return [wrapper, instance]
}

const renderApp = (props, customStore) => {
  const store = customStore || appStoreGen()
  const wrapper = render(
    <MockApp customStore={store}>
      <CellData {...props} />
    </MockApp>
  )
  return [wrapper, store]
}

describe('CellData', () => {
  afterEach(cleanup)

  describe('props', () => {
    test('location is required', () => {
      const props = {...requiredProps}
      delete props.location
      expect(() => createApp(props)).toThrow()
    })
  })

  describe('#evaluatedOnDoubleClick', () => {
    it('displays input field with current result value result tag is double-clicked', async () => {
      const { location } = testProps
      const entered = '5'
      const store = appStoreGen()
      await store.dispatch(setCellData(location, entered))
      renderApp(testProps, store)
      const resultTag = document.querySelector(`[data-location="${testProps.location}"]`)

      fireEvent.doubleClick(resultTag)
      const inputTag = document.querySelector(
        `[data-cell="input"][data-location="${testProps.location}"]`
        )
      expect(inputTag.value).toEqual(entered)
    })
  })

  describe('#evaluatedOnKeyDownEditable', () => {
    it('displays input field with empty field', async () => {
      const { location } = testProps
      const entered = 'test entered'
      const store = appStoreGen()
      await store.dispatch(setCellData(location, entered))
      renderApp(testProps, store)
      const tag = document.querySelector(`[data-location="${testProps.location}"]`)

      fireEvent.keyDown(tag, { key: 'l' })
      const inputTag = document.querySelector(
        `[data-cell="input"][data-location="${testProps.location}"]`
      )
      expect(inputTag.value).toEqual('')
    })
  })

  describe('#inputTagOnEscape', () => {
    it('dispalys focused text tag when escaping from input tag', () => {
      renderApp(testProps)

      const resultTag = document.querySelector(`[data-location="${testProps.location}"]`)
      fireEvent.doubleClick(resultTag)
      const inputTag = document.querySelector(
        `[data-cell="input"][data-location="${testProps.location}"]`
      )

      expect(document.activeElement).toBe(inputTag)
      fireEvent.keyDown(inputTag, { key: 'Escape' })
      expect(document.activeElement).toEqual(resultTag)
    })
  })

  describe('#inputTagOnCommit', () => {
    it('displays text tag with current text value and is not focused', () => {
      const value = 'test value'
      let resultTag
      renderApp(testProps)

      resultTag = document.querySelector(`[data-location="${testProps.location}"]`)
      fireEvent.doubleClick(resultTag)
      const inputTag = document.querySelector(
        `[data-cell="input"][data-location="${testProps.location}"]`
      )

      inputTag.value = value
      fireEvent.keyDown(inputTag, { key: 'Enter' })
      resultTag = document.querySelector(`[data-location="${testProps.location}"]`)
      expect(resultTag.innerHTML).toEqual(value)
      expect(document.activeElement).not.toBe(resultTag)
    })
  })

  describe('functional tests', () => {
    it('on component mount start state machine interpereter', () => {
      jest.spyOn(xstate.Interpreter.prototype, 'start')
      renderApp(testProps)
      expect(xstate.Interpreter.prototype.start).toHaveBeenCalledTimes(1)
    })

    it('on component unmount stop state machine interpereter', () => {
      jest.spyOn(xstate.Interpreter.prototype, 'stop')
      renderApp(testProps)
      cleanup()
      expect(xstate.Interpreter.prototype.stop).toHaveBeenCalledTimes(1)
    })

    test('lifecycle snapshots', () => {
      let resultTag, inputTag
      const [wrapper, _] = renderApp(testProps)
      expect(wrapper.asFragment()).toMatchSnapshot('1. initial text tag')

      resultTag = document.querySelector(`[data-location="${testProps.location}"]`)
      fireEvent.doubleClick(resultTag)
      expect(wrapper.asFragment()).toMatchSnapshot('2. initial input tag')

      inputTag = document.querySelector(
        `[data-cell="input"][data-location="${testProps.location}"]`
      )
      inputTag.value = '=2+3+4'
      fireEvent.keyDown(inputTag, { key: 'Enter' })
      expect(wrapper.asFragment()).toMatchSnapshot('3. text tag has text')

      resultTag = document.querySelector(`[data-location="${testProps.location}"]`)
      fireEvent.doubleClick(resultTag)
      expect(wrapper.asFragment()).toMatchSnapshot('4. input tag has formula')

      inputTag = document.querySelector(
        `[data-cell="input"][data-location="${testProps.location}"]`
      )
      fireEvent.keyDown(inputTag, { key: 'Escape' })
      resultTag = document.querySelector(`[data-location="${testProps.location}"]`)
      fireEvent.keyDown(resultTag, { key: 'Backspace' })
      expect(wrapper.asFragment()).toMatchSnapshot('5. cell value deleted with Backspace')

      fireEvent.keyDown(resultTag, { key: 'a' })
      inputTag = document.querySelector(
        `[data-cell="input"][data-location="${testProps.location}"]`
      )
      expect(wrapper.asFragment()).toMatchSnapshot('6. key down "a" on text tag')
    })
  })
})
