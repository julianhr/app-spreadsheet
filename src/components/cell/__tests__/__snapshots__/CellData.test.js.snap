// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellData functional tests lifecycle snapshots: 1. initial text tag 1`] = `
<DocumentFragment>
  <div
    class="row-header-height col-label-width css-1wpoxe-Wrapper eoyj2860"
  >
    <div
      class="css-16d5k2m-DataTag ekht17d0"
      data-cell="result"
      data-location="B-2"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;

exports[`CellData functional tests lifecycle snapshots: 2. initial input tag 1`] = `
<DocumentFragment>
  <div
    class="row-header-height col-label-width css-1wpoxe-Wrapper eoyj2860"
  >
    <input
      class="css-1obixtf-Input egik0ru0"
      data-cell="input"
      data-location="B-2"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`CellData functional tests lifecycle snapshots: 3. text tag has text 1`] = `
<DocumentFragment>
  <div
    class="row-header-height col-label-width css-1wpoxe-Wrapper eoyj2860"
  >
    <div
      class="css-11zzl5v-DataTag ekht17d0"
      data-cell="result"
      data-location="B-2"
      tabindex="0"
    >
      9
    </div>
  </div>
</DocumentFragment>
`;

exports[`CellData functional tests lifecycle snapshots: 4. input tag has formula 1`] = `
<DocumentFragment>
  <div
    class="row-header-height col-label-width css-1wpoxe-Wrapper eoyj2860"
  >
    <input
      class="css-1obixtf-Input egik0ru0"
      data-cell="input"
      data-location="B-2"
      value="=2+3+4"
    />
  </div>
</DocumentFragment>
`;

exports[`CellData functional tests lifecycle snapshots: 5. cell value deleted with Backspace 1`] = `
<DocumentFragment>
  <div
    class="row-header-height col-label-width css-1wpoxe-Wrapper eoyj2860"
  >
    <div
      class="css-16d5k2m-DataTag ekht17d0"
      data-cell="result"
      data-location="B-2"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;

exports[`CellData functional tests lifecycle snapshots: 6. key down "a" on text tag 1`] = `
<DocumentFragment>
  <div
    class="row-header-height col-label-width css-1wpoxe-Wrapper eoyj2860"
  >
    <input
      class="css-1obixtf-Input egik0ru0"
      data-cell="input"
      data-location="B-2"
      value=""
    />
  </div>
</DocumentFragment>
`;
